core/offchain/impl/offchain_worker_impl.cpp

```diff
@@ -0,0 +1,340 @@
+
+    return result;
+  }
+
+  Timestamp OffchainWorkerImpl::timestamp() {
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
+               clock_->now().time_since_epoch())
+        .count();
+  }
+
+  void OffchainWorkerImpl::sleepUntil(Timestamp deadline) {
+    auto ts = clock_->zero() + std::chrono::milliseconds(deadline);
+    SL_TRACE(log_,
+             "Falling asleep till {} (for {}ms)",
+             deadline,
+             std::chrono::duration_cast<std::chrono::milliseconds>(
+                 ts - clock_->now())
+                 .count());
+
+    std::this_thread::sleep_until(std::chrono::system_clock::time_point()
+                                  + std::chrono::milliseconds(deadline));
+    SL_DEBUG(log_, "Woke up after sleeping");
+  }
+
+  RandomSeed OffchainWorkerImpl::randomSeed() {
+    RandomSeed seed_bytes;
+    random_generator_->fillRandomly(seed_bytes);
+    return seed_bytes;
+  }
+
+  offchain::OffchainStorage &OffchainWorkerImpl::getStorage(
+      StorageType storage_type) {
+    switch (storage_type) {
+      case StorageType::Persistent:
+        return *persistent_storage_;
+
+      case StorageType::Local:
+        // TODO(xDimon):
+        //  Need to implemented as soon as it will implemented in Substrate.
+        //  Specification in not enough to implement it now.
+        SL_WARN(
+            log_,
+            "Attempt to use off-chain local storage which unavailable yet.");
+        return *local_storage_;
+
+      case StorageType::Undefined:
+      default:
+        BOOST_UNREACHABLE_RETURN({});
+    }
+  }
+
+  void OffchainWorkerImpl::localStorageSet(StorageType storage_type,
+                                           const common::Buffer &key,
+                                           common::Buffer value) {
+    auto &storage = getStorage(storage_type);
+    auto result = storage.set(key, std::move(value));
+    if (result.has_error()) {
+      SL_WARN(log_, "Can't set value in storage: {}", result.error().message());
+    }
+  }
+
+  void OffchainWorkerImpl::localStorageClear(StorageType storage_type,
+                                             const common::Buffer &key) {
+    auto &storage = getStorage(storage_type);
+    auto result = storage.clear(key);
+    if (result.has_error()) {
+      SL_WARN(
+          log_, "Can't clear value in storage: {}", result.error().message());
+    }
+  }
+
+  bool OffchainWorkerImpl::localStorageCompareAndSet(
+      StorageType storage_type,
+      const common::Buffer &key,
+      std::optional<std::reference_wrapper<const common::Buffer>> expected,
+      common::Buffer value) {
+    auto &storage = getStorage(storage_type);
+    auto result = storage.compare_and_set(key, expected, std::move(value));
+    if (result.has_error()) {
+      SL_WARN(log_,
+              "Can't compare-and-set value in storage: {}",
+              result.error().message());
+    }
+    return result.value();
```

Comment:
Is it supposed to throw an exception on error? Won't it be better to do it explicitly then?

Reply:
Maybe you are right. I'll research it

Reply:
Add missed return in condition